versioned_docs/version-1.5(WIP)/gwScripts.md

```diff
@@ -0,0 +1,204 @@
+
+## Layer 2 Scripts
+
+The C scripts that located in the `c` directory are Godwoken layer-2 scripts. The layer-2 script can be executed on CKB when a challenge occurs, which means that a layer-2 script is also a valid layer-1 script, only it follows the special interface convenience required by Godwoken.
+
+Godwoken account consisted of the following fields: `(id: u32, nonce: u32, script: Script)`. The `script` field determines the script that the account used. 
+
+Layer-2 scripts exist in two types: lock and contract. If an account ID appears in `l2tx.from_id`, we will assume that the account's script is a lock, implying that the script follows the lock script interface convenience that can verify signatures(e.g. Ethereum EOA). If an account ID appears in `l2tx.to_id`, we will assume that the account's script is a contract, which means we should execute the tx when it is sent to the account(like an Ethereum contract account).
+
+A layer-2 contract script is run both on and off-chain. The unified interface is defined in the `c/gw_def.h`. The on-chain implementation is `validator_utils.h`, and the off-chain implementation is `generator_utils.h`.
+
+### ETH Account Lock
+
+ETH Account Lock is a 2-layer lock script that verifies the layer 2 account signature.
+
+### Meta Contract
+
+Meta contract is a layer-2 contract script. A built-in layer-2 account allows creating another account by sending a tx to the account. Meta contract args is `MetaContractArgs`, the built-in contract id is `0`.
+
+### sUDT Contract
+
+sUDT contract is a layer-2 contract script that keeps a consistent mapping to the layer-1 sUDT. The layer-1 sUDT script hash is equal to `account.script.args`. When the user deposits a new type of sUDT, Godwoken will create a new corresponding sUDT account. This contract args is `SUDTArgs`, the built-in CKB Simple UDT contract id is `1`.
+
+### ETH Address Registry
+
+ETH address registry is a layer-2 contract that handles the mapping of Ethereum addresses to Godwoken accounts.
+
+When a user deposits token to create a new account, a corresponding Ethereum address will be inserted into the contract. If the account is created through a Meta contract, the user must register the Ethereum address for the acount by calling the ETH address registry contract.
+
+The built-in ETH address registry is allocated to id `2`.
+
+### [Polyjuice](https://github.com/nervosnetwork/godwoken-polyjuice)
+
+Polyjuice is a backend of Godwoken for state computation. The C scripts are in the `c` directory and are built using the command `make all-via-docker`. All tests run with the command `bash devtools/ci/integration-test.sh`.
+
+Polyjuice backend accepts an Ethrereum-like transaction and executes it in EVM. Ethereum [transaction structure](https://eth.wiki/json-rpc/API#eth_sendtransaction) is as follows:
```

Comment:
- in EVM --> in the EVM
- Ethereum [transaction structure] --> The Ethereum [transaction structure]